ros_bt_py/src/ros_bt_py/ros_nodes/action.py

```diff
@@ -164,106 +195,148 @@ def _do_setup(self):
+        self.logdebug(f"Received feedback message: {feedback}")
         with self._lock:
             self._feedback = feedback
 
     def _do_tick_wait_for_action_complete(self) -> str:
-        if self._running_goal_future is None:
-            self.logfatal("Result handle future is none! Cannot continue!")
+        if self._running_goal_handle is None or self._running_goal_future is None:
+            self._internal_state = ActionStates.FINISHED
             return NodeMsg.BROKEN
+
         if self._running_goal_future.done():
             result = self._running_goal_future.result()
             if result is None:
                 self._running_goal_handle = None
                 self._running_goal_future = None
                 self._active_goal = None
+
+                self._internal_state = ActionStates.FINISHED
+
+                self.logdebug("Action result is none, action call must have failed!")
                 return NodeMsg.FAILED
-            self.outputs["result"] = result
+
+            self.outputs["result"] = result.result
             self._running_goal_handle = None
             self._running_goal_future = None
             self._active_goal = None
+
+            self._internal_state = ActionStates.FINISHED
+
+            self.logdebug("Action succeeded, publishing result!")
             return NodeMsg.SUCCEED
+
         if self._running_goal_future.cancelled():
             self._running_goal_handle = None
             self._running_goal_future = None
             self._active_goal = None
+
+            self._internal_state = ActionStates.FINISHED
+
+            self.logwarn("Action execution was cancelled by the remote server!")
             return NodeMsg.FAILED
         seconds_running = (
             self._running_goal_start_time - self.ros_node.get_clock().now()
         ).nanoseconds / 1e9
+
         if seconds_running > self.options["timeout_seconds"]:
             self.logwarn(f"Cancelling goal after {seconds_running:f} seconds!")
             self._running_goal_future.cancel()
             self._running_goal_future = None
-            status = self._do_tick_cancel_running_goal()
-            if status != NodeMsg.SUCCEED:
-                return status
+            self._internal_state = ActionStates.REQUEST_GOAL_CANCELLATION
+            return NodeMsg.SUCCEED
+
         return NodeMsg.RUNNING
 
     def _do_tick_cancel_running_goal(self) -> str:
         if self._running_goal_handle is None:
-            self.logfatal("Result handle future is none! Cannot continue!")
+            self.logwarn(
+                "Goal cancellation was requested, but there is no handle to the running goal!"
+            )
+            self._internal_state = ActionStates.FINISHED
             return NodeMsg.BROKEN
+
         self._cancel_goal_future = self._running_goal_handle.cancel_goal_async()
-        return NodeMsg.RUNNING
+        self._internal_state = ActionStates.WAITING_FOR_GOAL_CANCELLATION
+        return NodeMsg.SUCCEED
```

Comment:
Same as above

Reply:
In fact this is done correctly here.
The goal is cancelled asynchronously, thus we need to check the result later (in the WAITING_FOR_GOAL_CANCELLATION state).
Secondly the SUCCESS status indicates that the cancellation request was made successfully.
This is not returned for the tick but used to determine if we need to evaluate another internal state or exit the entire tick prematurely:
https://github.com/fzi-forschungszentrum-informatik/ros2_ros_bt_py/blob/c30152f33d163fb66b7d6e2114cbe52f3578b8cc/ros_bt_py/src/ros_bt_py/ros_nodes/action.py#L377